### ISSUE TITLE
Example of connecting to IoT Central with x.509 certs in C language

### ISSUE THREAD

**Issue by InhalienableRights** [Azure-Samples/iot-central-high-availability-clients#2]
### This issue is for a:
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Desired behavior
> A simplified C language example of using DPS with x.509 leaf certificates, similar to the Java based [samples](https://docs.microsoft.com/en-us/azure/iot-central/core/how-to-connect-devices-x509?pivots=programming-language-java) where you can pull the cert/private key from PEM files or even pasted into the sample code. 

### OS and Version?
> Windows 10 or Linux (ubuntu 20.04).

### Versions
> IoT Central v3

### Details that might be useful
> The samples in the C SDK for PnP do work well with symmetric keys, though I have been getting stuck after setting cert/private key with IoTHubDeviceClient_LL_SetOption(). The server returns a mqtt_info object that has both x.509 cert/private key values looking correct but the server response (errorCode: 401002, CA certificate not found.) suggests I have something configured incorrectly. I do have the java sample working so it would seem the problem I am having is in the C client and not with the certs or server.

**Comment by firedog1024** [comment]
Thanks for the suggestion of adding X.509 support.  As you have noted it's bit of a pain to get the X.509 certs working and they add a fair amount of complexity to the code that we were looking to avoid as this was really a focus on how to get the High Availability and failover scenario working on Azure IoT Central.  Our thinking was that for alternative authentication methods folks would just use the standard SDK samples to see how to use the authentication method. Have you looked at the sample here: https://github.com/Azure/azure-iot-sdk-c/blob/main/provisioning_client/samples/prov_dev_client_ll_sample/prov_dev_client_ll_sample.c that shows using X.509 certs for DPS?

I also did a search for the error you are seeing (401002) on the C SDK GitHub issues and there are a number of posts on this issue with answers from the C SDK development team that might be useful to you and a faster way to get unblocked.  Failing finding a resolution from a previous issue I'd also suggest posting your problem on the C SDK issues as they are likely to be able to help much faster than we can.

Sorry I cannot provide a solution right away but I hope the resources I've listed above will be useful and help you resolve this issue.

**Comment by srkalava** [comment]
We will consider adding X.509 sample in the future, thanks for the suggestion!

**Comment by InhalienableRights** [comment]
Thank you for the response! 

I have played around with that sample however the issue I ran into, with that one, is that I need to be able to use a custom cert (a leaf cert generated from the root that is hosted in IoTCentral) and I could not seem to get that working. I think due to something signing based with the hsm...either way that same error code was happening, when passing my custom cert, even though with java, javascript, or C# everything works flawlessly. I am pretty inexperienced with c/c++ so that maybe playing a part.
I appreciate you looking into that error code, I have also looked there and tried a few of the suggestions to no avail. However, I have not thought to post there though so if i get completely frustrated ill give that a shot!
Thanks again,-Jesse

    On Friday, January 7, 2022, 03:17:06 PM PST, Ian Hollier ***@***.***> wrote:  
 
 


Thanks for the suggestion of adding X.509 support. As you have noted it's bit of a pain to get the X.509 certs working and they add a fair amount of complexity to the code that we were looking to avoid as this was really a focus on how to get the High Availability and failover scenario working on Azure IoT Central. Our thinking was that for alternative authentication methods folks would just use the standard SDK samples to see how to use the authentication method. Have you looked at the sample here: https://github.com/Azure/azure-iot-sdk-c/blob/main/provisioning_client/samples/prov_dev_client_ll_sample/prov_dev_client_ll_sample.c that shows using X.509 certs for DPS?

I also did a search for the error you are seeing (401002) on the C SDK GitHub issues and there are a number of posts on this issue with answers from the C SDK development team that might be useful to you and a faster way to get unblocked. Failing finding a resolution from a previous issue I'd also suggest posting your problem on the C SDK issues as they are likely to be able to help much faster than we can.

Sorry I cannot provide a solution right away but I hope the resources I've listed above will be useful and help you resolve this issue.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
